Alluxio/alluxio
# [ALLUXIO-3190] Show configuration errors and warnings on Web UI

https://alluxio.atlassian.net/browse/ALLUXIO-3190

# Test Guideline:
It is difficult to use unit tests or integration tests to test whether our web UI shows data as expected, so we would be better to manually test it.

Our web UI will get ```ConfigCheckReport``` from ```ServerConfigurationChecker```, process and show on the Web UI. We have unit tests to test the functionality of ```ServerConfigurationRecord``` and ```ServerConfigurationChecker```, so we can directly create ```ConfigCheckReport``` for different situations(PASSED, WARN, FAILED) and see if our website shows the data as what we want.

## Web UI shows errors and warnings:
For example, we can replace 
``` java
  /**
   * @return the configuration checker report
   */
  public synchronized ConfigCheckReport getConfigCheckReport() {
    return mConfigCheckReport;
  }
```
in ServerConfigurationChecker with
``` java
  /**
   * @return the configuration checker report
   */
  public synchronized ConfigCheckReport getConfigCheckReport() {
    // Randomly generated a ConfigCheckReport to test the functionality of Web UI
    List<WrongProperty> errors = new ArrayList<>();
    Map<String, List<String>> valuesA = new HashMap<>();
    valuesA.put("hdfs://name1:port1", Arrays.asList("masterHostname1:3123", "masterHostname2:6313"));
    valuesA.put("hdfs://name2:port2", Arrays.asList("masterHostname3:9412", "masterHostname4:431431"));
    Map<String, List<String>> valuesB = new HashMap<>();
    valuesB.put("ErrorValue1", Arrays.asList("masterHostname4:431431", "masterHostname2:6313"));
    valuesB.put("ErrorValue2", Arrays.asList("masterHostname1:3123", "masterHostname3:9412"));
    errors.add(new WrongProperty().setName("alluxio.property.key").setValues(valuesA));
    errors.add(new WrongProperty().setName("alluxio.configuration.key").setValues(valuesB));

    List<WrongProperty> serverWarns = new ArrayList<>();
    Map<String, List<String>> valuesC = new HashMap<>();
    valuesC.put("/a/b/c", Arrays.asList("workerHostname1:1321", "workerHostname2:2197"));
    valuesC.put("/d/e/f", Arrays.asList("masterHostname1:3123", "masterHostname2:6313"));
    serverWarns.add(new WrongProperty().setName("alluxio.property.key.second").setValues(valuesC));

    List<WrongProperty> workerWarns = new ArrayList<>();
    Map<String, List<String>> valuesD = new HashMap<>();
    valuesD.put("WarnValue1", Arrays.asList("workerHostname1:1321", "workerHostName3:4281"));
    valuesD.put("WarnValue2", Arrays.asList("workerHostName1:4831", "workerHostname2:2197"));
    workerWarns.add(new WrongProperty().setName("alluxio.configuration.key.second").setValues(valuesD));

    Map<Scope, List<WrongProperty>> errorsMap = new HashMap<>();
    Map<Scope, List<WrongProperty>> warnsMap = new HashMap<>();
    Map<Scope, Status> statusMap = new HashMap<>();

    errorsMap.put(Scope.MASTER, errors);
    warnsMap.put(Scope.SERVER, serverWarns);
    warnsMap.put(Scope.WORKER, workerWarns);
    statusMap.put(Scope.MASTER, Status.FAILED);
    statusMap.put(Scope.WORKER, Status.WARN);

    return new ConfigCheckReport(errorsMap, warnsMap, statusMap, Status.FAILED);
  }
```
Web UI can get the ConfigCheckReport created by  thoses codes 
and shows inconsistent properties and inconsistent properties details like:

![screen shot 2018-05-14 at 8 58 02 am](https://user-images.githubusercontent.com/25273483/40008911-60cde4d8-5755-11e8-9fa2-29e4fd6dc091.png)

## Web UI shows warnings:
Replace 
``` java
  /**
   * @return the configuration checker report
   */
  public synchronized ConfigCheckReport getConfigCheckReport() {
    return mConfigCheckReport;
  }
```
in ServerConfigurationChecker with
```java
  /**
   * @return the configuration checker report
   */
  public synchronized ConfigCheckReport getConfigCheckReport() {
    // Randomly generated a ConfigCheckReport to test the functionality of Web UI
    List<WrongProperty> serverWarns = new ArrayList<>();
    Map<String, List<String>> valuesC = new HashMap<>();
    valuesC.put("/a/b/c", Arrays.asList("workerHostname1:1321", "workerHostname2:2197"));
    valuesC.put("/d/e/f", Arrays.asList("masterHostname1:3123", "masterHostname2:6313"));
    serverWarns.add(new WrongProperty().setName("alluxio.property.key.second").setValues(valuesC));

    List<WrongProperty> workerWarns = new ArrayList<>();
    Map<String, List<String>> valuesD = new HashMap<>();
    valuesD.put("WarnValue1", Arrays.asList("workerHostname1:1321", "workerHostName3:4281"));
    valuesD.put("WarnValue2", Arrays.asList("workerHostName1:4831", "workerHostname2:2197"));
    workerWarns.add(new WrongProperty().setName("alluxio.configuration.key.second").setValues(valuesD));

    Map<Scope, List<WrongProperty>> errorsMap = new HashMap<>();
    Map<Scope, List<WrongProperty>> warnsMap = new HashMap<>();
    Map<Scope, Status> statusMap = new HashMap<>();

    warnsMap.put(Scope.SERVER, serverWarns);
    warnsMap.put(Scope.WORKER, workerWarns);
    statusMap.put(Scope.MASTER, Status.PASSED);
    statusMap.put(Scope.WORKER, Status.WARN);

    return new ConfigCheckReport(errorsMap, warnsMap, statusMap, Status.WARN);
  }

```
The Web UI is expected to show the warnings:
![screen shot 2018-05-14 at 4 46 51 pm](https://user-images.githubusercontent.com/25273483/40028955-a96269c6-5796-11e8-8523-48c5a3fa2682.png)

My own computer (show configuration warnings):
![screen shot 2018-05-14 at 9 11 13 am](https://user-images.githubusercontent.com/25273483/40009448-e0525b7a-5756-11e8-8cf3-921b9514a8f1.png)

## Web UI shows no errors/warnings:
Replaced with 
```java
  /**
   * @return the configuration checker report
   */
  public synchronized ConfigCheckReport getConfigCheckReport() {
    Map<Scope, Status> statusMap = new HashMap<>();
    statusMap.put(Scope.MASTER, Status.PASSED);
    statusMap.put(Scope.WORKER, Status.PASSED);
    return new ConfigCheckReport(new HashMap<>(), new HashMap<>(), statusMap, Status.PASSED);
  }
```
If ServerConfigCheck passed, no inconsistent properties num and details will be shown on the web UI.
![screen shot 2018-05-14 at 4 53 15 pm](https://user-images.githubusercontent.com/25273483/40029071-5c0a7c9e-5797-11e8-872b-1ff2607788f4.png)
